File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int getHeight(TreeNode* root, bool &isHeightBalanced) {
        if (root == NULL) return 0;

        int leftHeight = getHeight(root->left, isHeightBalanced);
        int rightHeight = getHeight(root->right, isHeightBalanced);

        if (abs(leftHeight - rightHeight) > 1) isHeightBalanced = false;

        return 1 + max(leftHeight, rightHeight);
    }
public:
    bool isBalanced(TreeNode* root) {
        bool isHeightBalanced = true;
        getHeight(root, isHeightBalanced);
        return isHeightBalanced;
    }
};

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isBalanced(root *TreeNode) bool {
    isHeightBalanced := true;
    getHeight(root, &isHeightBalanced)
    return isHeightBalanced
}

func getHeight(root *TreeNode, isHeightBalanced *bool) int {
    if root == nil {
        return 0
    }

    left := getHeight(root.Left, isHeightBalanced)
    right := getHeight(root.Right, isHeightBalanced)

    if math.Abs(float64(left) - float64(right)) > 1 {
        *isHeightBalanced = false
    }

    return 1 + max(left, right)
}